neuralop/training/callbacks.py

```diff
@@ -375,82 +376,123 @@ def on_before_loss(self, out):
+        save_dir : Union[Path, str], optional
+            folder in which to save checkpoints, by default './checkpoints'
+        save_best : str, optional
+            metric to monitor for best value in order to save state
+        save_step : int, optional
+            interval on which to save/check metric, by default 1
+        save_optimizer : bool, optional
+            whether to save optimizer state, by default False
+        save_scheduler : bool, optional
+            whether to save scheduler state, by default False
+        save_regularizer : bool, optional
+            whether to save regularizer state, by default False
+        resume_from_dir : Union[Path, str], optional
+            folder from which to resume training state. 
+            Expects saved states in the form: (all but model optional)
+                model.pt, optimizer.pt, scheduler.pt, regularizer.pt
+            All state files present will be loaded. 
         """
+        
         super().__init__()
-
-        if isinstance(checkpoint_dir, str):
-            checkpoint_dir = Path(checkpoint_dir)
-
-        if not checkpoint_dir.exists():
-            checkpoint_dir.mkdir(parents=True)
-        self.checkpoint_dir = checkpoint_dir
-        self.interval = interval
+        if isinstance(save_dir, str): 
+            save_dir = Path(save_dir)
+        if not save_dir.exists():
+            save_dir.mkdir(parents=True)
+        self.save_dir = save_dir
+
+        self.save_step = save_step
+        self.save_best = save_best
+        self.save_optimizer = save_optimizer
+        self.save_scheduler = save_scheduler
+        self.save_regularizer = save_regularizer
+
+        if resume_from_dir:
+            if isinstance(resume_from_dir, str):
+                resume_from_dir = Path(resume_from_dir)
+            assert resume_from_dir.exists()
+
+        self.resume_from_dir = resume_from_dir
+        
 
     def on_init_end(self, *args, **kwargs):
         self._update_state_dict(**kwargs)
+
     
-    def on_epoch_start(self, *args, **kwargs):
+    def on_train_start(self, *args, **kwargs):
         self._update_state_dict(**kwargs)
 
-    def on_epoch_end(self, *args, **kwargs):
-        if self.state_dict['epoch'] % self.interval == 0:
-            checkpoint_path = self.checkpoint_dir / f"ep_{self.state_dict['epoch']}.pt"
-            torch.save(self.state_dict['model'].state_dict(), checkpoint_path)
+        if self.save_best:
+            assert self.state_dict['eval_losses'], "Error: cannot monitor a metric if no validation metrics exist."
+            assert self.save_best in self.state_dict['eval_losses'].keys(), "Error: cannot monitor a metric outside of eval_losses."
+            self.best_metric_value = float['inf']
+        else:
+            self.best_metric_value = None
         
+        # load state dict if resume_from_dir is given
+        if self.resume_from_dir:
+            save_fnames = os.listdir(self.resume_from_dir)
```

Comment:
Let's not use os, it is not portable. Let's used pathlib.Path and directly glob or better just joinpath with the files we need